### PR TITLE
New cache file linking guide and related updates to existing docs

### PR DESCRIPTION
**Pull request by jorgeorpinel**
Partially addresses #265

### CODE REVIEW

**Comment by jorgeorpinel** [comment]
@shcheklein the guide isn't here yet but you may want to double check the file changes in this PR so far (reflecting https://github.com/iterative/dvc/pull/1841), which are needed for consistency with recent updates I made (#344) and with the new user guides I'll push up here soon. Thanks ✌️

File: static/docs/get-started/add-files.md [hunk 1]
```diff
@@ -41,7 +41,7 @@ $ git commit -m "add source data to DVC"
 ### Expand to learn about DVC internals
 
 You can see that actual data file has been moved to the `.dvc/cache` directory
-(usually hardlink or reflink is created, so no physical copying is happening).
+(usually a reflink is created, so no file content copying is happening).
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply static/docs/get-started/add-files.md]: I would actually make an details section - something like: `IMPORTANT: Performance on large data or something` ...

File: static/docs/get-started/example-pipeline.md [hunk 2]
```diff
@@ -167,8 +167,8 @@ intermediate result.
 Second, you should see by now that the actual data is stored in the `.dvc/cache`
 directory, each file having a name in a form of an md5 hash. This cache is
 similar to Git's internal objects store but made specifically to handle large
-data files. DVC is using reflinks, hardlinks and other optimizations to manage
-your actual workspace without copying every time object from/to the cache.
+data files. DVC uses reflinks, hardlinks, or other optimizations to manage your
```
Comment (shcheklein) [inline static/docs/get-started/example-pipeline.md]: the same, we should be emphasizing this. It was really painful decision to disable hardlinks and symlinks. A lot of ppl might be confused and disappointed to see that we are copying their files. We should give a lot of warnings and docs so that they understand right away that there is a way to enable an optimization.

File: static/docs/philosophy/related-technologies.md [hunk 3]
```diff
@@ -81,7 +81,8 @@ process.
      want to see in your Git repository) in a local key-value store and use file
      symlinks instead of the actual files.
 
-   - DVC uses hardlinks instead of symlinks to make user experience better.
+   - DVC uses reflinks (or hardlinks) instead of symlinks to make user
```
Comment (shcheklein) [inline static/docs/philosophy/related-technologies.md]: I think the previous paragraph should be updated. I don't see how two of them make sense together. @dmpetrov @efiop what did we want to actually say here? are you familiar with git-annex internals?

File: static/docs/philosophy/related-technologies.md [hunk 4]
```diff
@@ -111,10 +112,9 @@ process.
      `git clone` command. It gives more granularity on managing data and code
      separately. Hooks could be configured to make workflow simpler.
 
-   - DVC creates hardlinks (or even reflinks if they are supported). The
-     `dvc checkout` command does not actually copy data files from cache to the
-     workspace, as copying files is a heavy operation for large files (30
-     GB+).
+   - DVC creates reflinks. The `dvc checkout` command does not actually copy
```
Comment (shcheklein) [inline static/docs/philosophy/related-technologies.md]: reflinks are rare. This phrase could be misleading. Btw, in places like this I would use a term copy-on-write links (aka reflinks)  and would put a link to some docs about them. It's not a very well known stuff.

File: static/docs/philosophy/related-technologies.md (hunk 3 above)
Comment (efiop) [reply static/docs/philosophy/related-technologies.md]: @shcheklein Ah, right, this is an old one. Not sure why the first paragraph is that way. I would probably remove it for now.

Also, would rephrase a little: `DVC can use reflinks or hardlinks instead of symlinks to make user experience better.`

File: static/docs/get-started/add-files.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply static/docs/get-started/add-files.md]: OK. I added that tradeoff information to the new file link type guide (now pushed) and linked to it here instead.

File: static/docs/get-started/example-pipeline.md (hunk 2 above)
Comment (jorgeorpinel) [reply static/docs/get-started/example-pipeline.md]: Split into special note.

File: static/docs/philosophy/related-technologies.md (hunk 3 above)
Comment (jorgeorpinel) [reply static/docs/philosophy/related-technologies.md]: Updated.

File: static/docs/philosophy/related-technologies.md (hunk 4 above)
Comment (jorgeorpinel) [reply static/docs/philosophy/related-technologies.md]: Added as a foot note in this document but what would be a good source to link to about it? I haven't found anything that looks authoritative enough. Wikipedia? https://en.wikipedia.org/wiki/Copy-on-write

**Comment by jorgeorpinel** [comment]
OK guys I addressed the feedback but also had to push up all the other changes I have which add the new /doc/user-guide/cache-file-linking guide (See it [live in review app](https://dvc-org-pr-345.herokuapp.com/doc/user-guide/cache-file-linking)). Hopefully it's not too many changes to review now!

File: static/docs/user-guide/cache-file-linking.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
+# File Link Types for the DVC Cache
+
+File links are entries in the file system that don't necessarily hold the file
+contents, but which point to where the file is actually stored. DVC uses file
+links in your project's workspace to avoid copying them from/to the DVC cache.
```
Comment (shcheklein) [inline static/docs/user-guide/cache-file-linking.md]: them - sounds like "to avoid copying links"

File: static/docs/user-guide/cache-file-linking.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+> [ReFS](https://docs.microsoft.com/en-us/windows-server/storage/refs/refs-overview)
+> file systems.
+
+## Supported file links types and their tradeoffs
+
+There are pros and cons to the 3 supported link types (`reflink`, `hardlink`,
+`symlink` or soft link). While reflinks have all the benefits and none of the
+worries, they're not commonly supported in most platforms yet. Hard/soft links
+also optimize speed and space in the file system, but carry the risk of breaking
+your workflow ,since updating tracked files in the workspace causes data
```
Comment (shcheklein) [inline static/docs/user-guide/cache-file-linking.md]: ` ,` -> `, `

File: static/docs/commands-reference/config.md [hunk 7]
```diff
@@ -103,58 +103,29 @@ files](/doc/user-guide/dvc-files-and-directories) for more details.)
...
-     repo and your cache directory are located on different filesystems/drives
-     (i.e. repo is located on ssd for performance, but cache dir is located on
-     hdd for bigger storage).  
-     Please note that data file linked with symlink should never be edited in
-     place, but instead deleted and then replaced with a new file, otherwise it
-     might cause cache corruption and automatic deletion of a cache file by dvc.
-
-  4. **`copy`** - The most inefficient link type, yet the most widely supported
-     for any repo/cache FS combination. Suitable for scenarios with relatively
-     small data files, where copying them is not a performance/storage concern.
```
Comment (shcheklein) [inline static/docs/commands-reference/config.md]: ❤️

File: static/docs/get-started/example-pipeline.md [hunk 8]
```diff
@@ -167,8 +167,12 @@ intermediate result.
 Second, you should see by now that the actual data is stored in the `.dvc/cache`
 directory, each file having a name in a form of an md5 hash. This cache is
 similar to Git's internal objects store but made specifically to handle large
-data files. DVC is using reflinks, hardlinks and other optimizations to manage
-your actual workspace without copying every time object from/to the cache.
+data files.
+
+> **Note!** For performance with large data files, DVC can use file links from
+the cache to the workspace to avoid copying actual file contents. Refer to
```
Comment (shcheklein) [inline static/docs/get-started/example-pipeline.md]: did you run the pre-commit hook we have now? I think it puts `>` before each line

File: static/docs/get-started/example-versioning.md [hunk 9]
```diff
@@ -227,9 +227,10 @@ $ python train.py
 $ dvc add model.h5
 ```
 
-Note! `dvc remove` or `dvc unprotect` is required, otherwise `python train.py`
-will overwrite the existing file and may corrupt the cached version. Check this
-[guide](/doc/user-guide/update-tracked-file) to learn more.
+> **Note!** `dvc remove` or `dvc unprotect` is required, otherwise `python
```
Comment (shcheklein) [inline static/docs/get-started/example-versioning.md]: it's not required anymore. It's required only in the protected mode. By default it's disabled and it's safe to modify/overwrite files.

File: static/docs/get-started/add-files.md [hunk 10]
```diff
@@ -41,7 +41,9 @@ $ git commit -m "add source data to DVC"
 ### Expand to learn about DVC internals
 
 You can see that actual data file has been moved to the `.dvc/cache` directory
-(usually hardlink or reflink is created, so no physical copying is happening).
+(ideally with reflinks if available on the system, otherwise by file copy – see
```
Comment (shcheklein) [inline static/docs/get-started/add-files.md]: let's make it a separate `<details>` section - something like `Important performance caveats` or something like this.

File: static/docs/user-guide/update-tracked-file.md [hunk 11]
```diff
@@ -1,10 +1,14 @@
...
 Due to the way DVC handles linking between the data files in the cache and their
-counterparts in the working directory (see
-[#799](https://github.com/iterative/dvc/issues/799) and
-[#599](https://github.com/iterative/dvc/issues/599) for example), updating
-tracked files has to be carried out with caution.
+counterparts in the working directory (refer to [Cache File
+Linking](/docs/user-guide/cache-file-linking)), updating tracked files has to be
+carried out with caution to avoid data corruption when the DVC config option
+`cache.type` is set to `hardlink` or/and `symlink`. (See `dvc config cache` for
+more details on setting the cache file link types.)
```
Comment (shcheklein) [inline static/docs/user-guide/update-tracked-file.md]: is it `).`?

File: static/docs/user-guide/cache-file-linking.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
+# File Link Types for the DVC Cache
```
Comment (shcheklein) [inline static/docs/user-guide/cache-file-linking.md]: I'm thinking if we should rename it to more user friendlier version - "Performance with large data" or something similar? Or are you thinking to put together a separate sections for that?

File: static/docs/philosophy/related-technologies.md (hunk 4 above)
Comment (jorgeorpinel) [reply static/docs/philosophy/related-technologies.md]: Link added.

File: static/docs/user-guide/cache-file-linking.md (hunk 5 above)
Comment (jorgeorpinel) [reply static/docs/user-guide/cache-file-linking.md]: Changed to "avoid duplicating contents that will exist in the DVC cache."

File: static/docs/get-started/example-pipeline.md (hunk 8 above)
Comment (jorgeorpinel) [reply static/docs/get-started/example-pipeline.md]: I had not but just installed the pre-commit hook (which uses `prettier`) and it does add them indeed. BTW Rewrap also puts `>` before each line but only if a long comment has no line breaks yet, otherwise it doesn't...

Also, `prettier` re-wraps my paragraphs differently than VSCode Rewrap. (The pre-commit hook doesn't line-break links.)

So there's inconsistencies throughout the docs with these 2 things at least. Maybe I should just run `prettier` on everything?

Comment (jorgeorpinel) [reply static/docs/get-started/example-pipeline.md]: OK... I just ran ``npm bin`/prettier --write --prose-wrap=always` on all the files altered by this PR and the funny thing is that when `pre-commit` ran prettier again the results were different. So I'm not sure how to run `prettier` so it has the same settings as when it is ran by `pre-commit`.

Anyway, I noticed some problems with the auto-reformatting like that it changes how we wrote sidebar.json and it inserts more spaces to indent `yaml` code-blocks showing DVC-file examples so I'm turning it off for now. I'll ask about all this on Discord.

File: static/docs/user-guide/update-tracked-file.md (hunk 11 above)
Comment (jorgeorpinel) [reply static/docs/user-guide/update-tracked-file.md]: In this case the entire sentence is inside the parenthesis so `.)` is correct 😃

I double checked just now and found several sources verifying this e.g. https://www.quickanddirtytips.com/education/grammar/periods-and-parentheses

```
If the words inside the parentheses are a complete sentence, the period, question
mark, or exclamation point that ends the sentence goes inside the parenthesis:

Bring chocolate. (Squiggly likes sweets.)
```

File: static/docs/user-guide/cache-file-linking.md (hunk 12 above)
Comment (jorgeorpinel) [reply static/docs/user-guide/cache-file-linking.md]: Yes! Actually this is not even the main new guide you requested in #265 **add a user guide page about the optimization (advanced mode)**. My plan was to add another guide for that matter, which will link back to this one.

Maybe I've added too much optimization info already to this one guide so no longer sure what will the optimization guide will contain. Is there anything else to optimize besides cache file link types?

File: static/docs/get-started/example-versioning.md (hunk 9 above)
Comment (jorgeorpinel) [reply static/docs/get-started/example-versioning.md]: Good point. Note removed.

File: static/docs/get-started/add-files.md (hunk 10 above)
Comment (jorgeorpinel) [reply static/docs/get-started/add-files.md]: Ok, done. This doc may have too many expandable detail sections now though, you may want to see it in the review app: https://dvc-org-pr-345.herokuapp.com/doc/get-started/add-files

File: static/docs/commands-reference/config.md [hunk 13]
```diff
@@ -87,74 +87,45 @@ files](/doc/user-guide/dvc-files-and-directories) for more details.)
...
   > See also helper command `dvc cache dir` that properly transform paths
-  relative to the present working directory into relative to the project config
-  file.
+  > relative to the present working directory into relative to the project
+  > config file.
 
 - `cache.protected` - makes files in the workspace read-only. Possible values
   are `true` or `false` (default). Run `dvc checkout` for the change go into
-  effect. (It affects only files that are under DVC control.)
+  effect. (It affects only files that are under DVC control.)
```
Comment (shcheklein) [inline static/docs/commands-reference/config.md]: remove trailing whitespaces

File: static/docs/commands-reference/config.md [hunk 14]
```diff
@@ -87,74 +87,45 @@ files](/doc/user-guide/dvc-files-and-directories) for more details.)
...
 - `cache.protected` - makes files in the workspace read-only. Possible values
   are `true` or `false` (default). Run `dvc checkout` for the change go into
-  effect. (It affects only files that are under DVC control.)
+  effect. (It affects only files that are under DVC control.)  
   Due to the way DVC handles linking between the data files in the cache and
   their counterparts in the working directory, it's easy to accidentally corrupt
   the cached version of a file by editing or overwriting it. Turning this config
   option on forces you to run `dvc unprotect` before updating a file, providing
-  an additional layer of security to your data.
+  an additional layer of security to your data.
```
Comment (shcheklein) [inline static/docs/commands-reference/config.md]: trailing whitespaces

File: static/docs/user-guide/cache-file-linking.md (hunk 12 above)
Comment (shcheklein) [reply static/docs/user-guide/cache-file-linking.md]: no, not that much .. I think this guide alone is probably enough. I would rename, may be write a brief motivation - that space and speed are important and we can't just copy, etc, etc. and call it a section :)

File: static/docs/commands-reference/config.md (hunk 13 above)
Comment (jorgeorpinel) [reply static/docs/commands-reference/config.md]: They are in on purpose. In markdown they help to insert a line break within a paragraph.

Explained here: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#lists

File: static/docs/commands-reference/config.md (hunk 14 above)
Comment (jorgeorpinel) [reply static/docs/commands-reference/config.md]: It's another line break. Should I remove these and add full paragraphs breaks instead?

**Comment by jorgeorpinel** [comment]
Ready, @shcheklein

**Comment by jorgeorpinel** [comment]
Ready, @shcheklein. Just a couple discussions still open above for a final decision on your part please. Thanks

File: static/docs/commands-reference/config.md (hunk 14 above)
Comment (shcheklein) [reply static/docs/commands-reference/config.md]: Ah, np then. I'm totally fine with two paragraph when you think it's appropriate.

File: static/docs/commands-reference/config.md (hunk 13 above)
Comment (shcheklein) [reply static/docs/commands-reference/config.md]: So, is this an actual whitespace? I'm fine with that but I would expect it to be fragile - some editors remove trailing whitespaces on save and prettier won't return them most likely. Not a big deal though, let's not worry about this now.

**Comment by shcheklein** [comment]
Great stuff, @jorgeorpinel ! 🎉

File: static/docs/commands-reference/config.md (hunk 13 above)
Comment (jorgeorpinel) [reply static/docs/commands-reference/config.md]: You're right. We may want to avoid these in the future... I'll search all the docs to see if there's many occurrences.
